core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/score/director/AbstractScoreDirector.java

```diff
@@ -458,7 +458,16 @@ public void afterProblemPropertyChanged(Object problemFactOrEntity) {
         if (isConstraintConfiguration(problemFactOrEntity)) {
             setWorkingSolution(workingSolution); // Nuke everything and recalculate, constraint weights have changed.
         } else {
-            variableListenerSupport.resetWorkingSolution(); // TODO do not nuke the variable listeners
+            EntityDescriptor<Solution_> entityDescriptor =
+                    getSolutionDescriptor().findEntityDescriptor(problemFactOrEntity.getClass());
+            if (entityDescriptor == null) { // Problem fact changed.
+                variableListenerSupport.resetWorkingSolution(); // TODO do not nuke the variable listeners
```

Comment:
There is a massive optimization possible here. But I need to speak to Jirka as to how to make it happen.